### ISSUE TITLE
[Feature request][UI] Improve naming in "Area selection" tab for power samples for better UX

### ISSUE THREAD

**Issue by cphlipot1** [google/perfetto#910]
If i click and drag over power samples in the UI i see something that gives something that looks like the following:

![Image](https://github.com/user-attachments/assets/48492bd1-729e-4b41-819b-7f0aa8322f28)

I think there is a lot of useful data in that tab, but it is difficult for many users to decipher who don't already have an understanding of Perfetto internals relating to how the samples are stored and processed internally. To improve UX i suggest renaming some columns and removing others in the selection window to improve UX as follows:
- "**Name**" -> "**Power Rail**"
- "**Delta value**" -> "**Energy (uJ)**" or "**Energy (uWS)**"
- "**Rate /s**" -> "**Avg Power (uW)**"
- "Weighted avg value" -> (Remove this. it isn't really clear the applicability of averaging energy samples and due to the name may people misinterpret this as the average power instead of average energy)
- "**Count**" - > "**Number of Samples**"
- "**First Value**" -> (Remove this)
- "**Last Value**" -> (Remove this)
- "**Min Value**" -> (Remove this)
- "**Max Value**" -> (Remove this)

**Comment by LalitMaganti** [comment]
Yeah I think what is going on here is that you are seeing the generic view for aggregating any "counter like" data there, irrespective of the source etc.

We're in the process of totally rewriting area selection and I think customizing the information in the generic area selection (and/or adding custom area selection for things we care a lot about) makes a lot of sense.

@stevegolton FYI feedback.

**Comment by cphlipot1** [comment]
@LalitMaganti  I wanted to check if this rewrite you mentioned complete, and the code is now in a state where implementing this would be feasible?

**Comment by LalitMaganti** [comment]
@stevegolton is the right person to comment here.

**Comment by stevegolton** [comment]
I think the architecture is flexible enough to add this feature now, though there isn't necessarily a well lit path for it.

See how the generic counter aggregation panel is added right now:
https://github.com/google/perfetto/blob/main/ui/src/plugins/dev.perfetto.GenericAggregations/index.ts#L37

Essentially you just need to register a new area selection panel using `ctx.selection.registerAreaSelectionTab()` from a plugin, passing in an object that takes the current area selection object and renders the content (mithril vnodes).

Here's a noddy example....

```ts
ctx.selection.registerAreaSelectionTab({
  id: 'my-aggregation',
  name: 'My Aggregation',
  content: (selection) => {
    if (tracksMatch(selection)) {
      return renderAggregationPanel(selection);
    } else {
      return undefined; 
    }
  },
})
```

You'll find that most aggregation panels are in fact using the `createAggregationToTabAdaptor` adapter to adapt the older aggregators to the new interface, so creating a new aggregator for power tracks might be the best way to do this.

```ts
ctx.selection.registerAreaSelectionTab(
  createAggregationToTabAdaptor(ctx, new PowerCounterSelectionAggregator()),
);

// ~~snip~~

class PowerCounterSelectionAggregator implements AreaSelectionAggregator {
  // ...
}
```

Take a look at the existing aggregator to get an idea of how it might work.

Another issue is finding out which tracks to show this panel for. I'm not sure the best way to do this. Possibly we could add the track type to the track tags in the `dev.perfetto.TraceProcessorTrack` plugin and filter on that.

I don't know if you were volunteering to do the work, but we're always happy to accept PRs. It might be quite a while before we get to this feature otherwise.

**Comment by cphlipot1** [comment]
Thanks for the detailed response

> Essentially you just need to register a new area selection panel using ctx.selection.registerAreaSelectionTab() from a plugin, passing in an object that takes the current area selection object and renders the content (mithril vnodes).

I there are things already registered for generic aggregations. Will doing this just run my new aggregation, or will it run both?

> Another issue is finding out which tracks to show this panel for. I'm not sure the best way to do this. Possibly we could add the track type to the track tags in the dev.perfetto.TraceProcessorTrack plugin and filter on that.

are we running SQL queries at this stage to get this info on the track or is this expected to be in some typescript data structure? I have a pretty good idea for what tracks this needs to be applied to and i think there is enough info in the tables at least, both in terms of track name and track type that could be used to filter this.

> I don't know if you were volunteering to do the work, but we're always happy to accept PRs. It might be quite a while before we get to this feature otherwise.

Looking at your above explanation, i think i could look into this assuming there aren't any unexpected complication, although i can't commit to a specific timeline for adding this either.

**Comment by stevegolton** [comment]
> Will doing this just run my new aggregation, or will it run both?

It'll run both, that is unless you also change the generic aggregator to avoid aggregating on the power tracks as well (please don't add anything power-specific there).

I'd recommend leaving it and using the priority property to make your aggregation appear in front of others. The highest priority metric will appear by default.

> are we running SQL queries at this stage to get this info on the track or is this expected to be in some typescript data structure? I have a pretty good idea for what tracks this needs to be applied to and i think there is enough info in the tables at least, both in terms of track name and track type that could be used to filter this.

You'll have to run sql queries - tracks rarely if ever load all the data into JavaScript, only just enough data to be able to plot the graph. I think you should have the data you need from the track_ids stored in the track tags property + the time range of the area selection.

> Looking at your above explanation, i think i could look into this assuming there aren't any unexpected complication, although i can't commit to a specific timeline for adding this either.

Of course! I'd like to clarify that I haven't had a go at this yet so there could be issues I've overlooked. Happy discuss further if you run into any issues.

**Comment by cphlipot1** [comment]
> I'd recommend leaving it and using the priority property to make your aggregation appear in front of others. The highest priority tab will appear by default.

Do you know if this will still run the generic aggregation query? Some of the traces i have contain millions of samples and this generic aggregation query can already take 1+ seconds to run, so from a UI responsiveness perspective i would like to be able to avoid running that generic aggregation query when it is not needed.

**Comment by stevegolton** [comment]
Yeah that's a good point. It'll still run the query the moment you make the selection.

I'm not sure the best way to fix this.

Let's focus on getting it working first, then deal with filtering out unwanted aggregations after.

**Comment by LalitMaganti** [comment]
> Do you know if this will still run the generic aggregation query? Some of the traces i have contain millions of samples and this generic aggregation query can already take 1+ seconds to run

FWIW we should fix this with a combination of making the query faster and multi-threading in trace processor (working on some things to unblock that right now).

**Comment by cphlipot1** [comment]
> FWIW we should fix this with a combination of making the query faster and multi-threading in trace processor (working on some things to unblock that right now).

I agree we should try to optimize the query, but being able to omit it would be better is possible

If we adopt the original suggestion in this issue it would involve dropping "Weighted avg value" from the aggregation view in the new query, which would allow significant opportunities to simplify and optimize the query. Last time i looked into the query, it looked lot of compute is being spent on that "Weighted avg value" aggregation specifically which isn't particularly meaningful for power rails.

**Comment by cphlipot1** [comment]
@stevegolton  I think you've seen my recent PR already.

While i would like to merge that, one follow up change i would like to make would be to change the power sample aggregation to be selected by default. Looking through the code i see there is a priority that could control this, but i didn't see an obvious way to control this within the plugin: https://github.com/google/perfetto/blob/5457622ee9b725a7ef3b51f030fadd9fce424c7c/ui/src/frontend/viewer_page/current_selection_tab.ts#L106-L108

do you have any suggestions on if/how that could be done, or if there are any fundemental objections to making the power rail aggregation appear by default when aggregating over power rails?

**Comment by stevegolton** [comment]
Take a look in `createAggregationToTabAdaptor` and you'll find that the priority is set based on the selectors in the aggregator.

There's no well it path to tweak this priority when using the adapter, but perhaps the best option is to just have an additional optional argument which overrides the priority, then set this to some high value (200 ish) for the power aggregation tab.

> or if there are any fundemental objections to making the power rail aggregation appear by default when aggregating over power rails?

No it makes sense that the power one takes precedence when a power track is selected. The more specific aggregations should show first.

**Comment by LalitMaganti** [comment]
I think we can close this out now right? We have good aggregation for power rails AFAIK.
